exonum-java-binding/core/rust/src/proxy/runtime.rs

```diff
@@ -0,0 +1,624 @@
+        JavaRuntimeProxy {
+            exec: executor,
+            runtime_adapter: adapter,
+            deployed_artifacts: HashSet::new(),
+            started_services: HashMap::new(),
+            started_services_by_name: HashMap::new(),
+        }
+    }
+
+    fn parse_artifact(&self, artifact: &ArtifactId) -> Result<JavaArtifactId, ExecutionError> {
+        if artifact.runtime_id != Self::RUNTIME_ID as u32 {
+            Err(Error::IncorrectArtifactId.into())
+        } else {
+            artifact
+                .name
+                .parse()
+                .map_err(|_| Error::IncorrectArtifactId.into())
+        }
+    }
+
+    fn can_deploy_artifact(&self, id: &JavaArtifactId) -> Result<(), ExecutionError> {
+        if self.deployed_artifacts.contains(id) {
+            Err(Error::AlreadyDeployed.into())
+        } else {
+            Ok(())
+        }
+    }
+
+    fn add_deployed_artifact(&mut self, id: JavaArtifactId) {
+        self.deployed_artifacts.insert(id);
+    }
+
+    fn can_start_service(&self, instance: &Instance) -> Result<(), ExecutionError> {
+        if self.started_services.contains_key(&instance.id) {
+            Err(Error::AlreadyStarted.into())
+        } else {
+            Ok(())
+        }
+    }
+
+    fn add_started_service(&mut self, instance: Instance) {
+        self.started_services_by_name
+            .insert(instance.name.clone(), instance.id);
+        self.started_services.insert(instance.id, instance);
+    }
+
+    fn remove_started_service(&mut self, id: &InstanceId) {
+        let service = self.started_services.remove(id);
+        if let Some(instance) = service {
+            self.started_services_by_name.remove(&instance.name);
+        }
+    }
+
+    fn parse_jni<T>(res: JniResult<T>) -> Result<T, ExecutionError> {
+        res.map_err(|err| match err.0 {
+            JniErrorKind::JavaException => Error::JavaException.into(),
+            _ => Error::UnspecifiedError.into(),
+        })
+    }
+}
+
+impl fmt::Debug for JavaRuntimeProxy {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "JavaRuntimeProxy()")
+    }
+}
+
+impl From<Error> for ExecutionError {
+    fn from(value: Error) -> ExecutionError {
+        ExecutionError::new(
+            ErrorKind::runtime(value as u8),
+            format!("{:?}", value.clone()),
+        )
+    }
+}
+
+impl Runtime for JavaRuntimeProxy {
+    fn deploy_artifact(
+        &mut self,
+        artifact: ArtifactId,
+        deploy_spec: Any,
+    ) -> Box<dyn Future<Item = (), Error = ExecutionError>> {
+
+        let id = match self.parse_artifact(&artifact) {
+            Ok(id) => id,
+            Err(err) => return Box::new(Err(err).into_future()),
+        };
+
+        if let Err(err) = self.can_deploy_artifact(&id) {
+            return Box::new(Err(err).into_future());
+        }
+
+        Box::new(Self::parse_jni(self.exec.with_attached(|env| {
```

Comment:
(here and each other method) I don't see the exception handling. If it does occur, the native is responsible to act on it (handle, according to the Runtime method specification, and clear, so that it does not remain 'thrown').

Reply:
See also https://docs.oracle.com/en/java/javase/11/docs/specs/jni/design.html#java-exceptions , but remember that jni-rs detects exceptions for you.

Existing proxies do that, but they might also do something specific atop of it (not necessarily needed here).